src/misc/guillotine.ts

```diff
@@ -0,0 +1,235 @@
+  if (armorStand.hasGravity()) return; // Has gravity -> Blade already falling.
+  // Begin drop
+  armorStand.setGravity(true);
+  armorStand.setInvulnerable(true);
+  let previousY = armorStand.velocity.y;
+
+  // Blade falling
+  for (let i = 0; i < 200; i++) {
+    await wait(2, 'ticks');
+    const currentY = armorStand.velocity.y;
+    if (!armorStand.isValid() || currentY >= previousY) break;
+    previousY = currentY;
+    chopIfHit(armorStand.location, Math.abs(currentY));
+  }
+
+  // Blade hit ground
+  armorStand.world.playSound(
+    armorStand.location,
+    LAND_SOUND,
+    SoundCategory.BLOCKS,
+    0.6,
+    0.8,
+  );
+  armorStand.setGravity(false);
+  armorStand.setInvulnerable(false);
+}
+
+async function chopIfHit(location: Location, y_velocity: number) {
+  const entities = location.getNearbyEntities(1, 1.2, 1);
+  for (const entity of entities) {
+    if (!(entity instanceof LivingEntity)) return;
+    if (entity instanceof ArmorStand) return;
+    // Set cooldown
+    if (cooldowns.has(entity)) return;
+    cooldowns.add(entity);
+
+    // Wait for better timing
+    await wait(4, 'ticks');
+
+    // Deal damage
+    if (y_velocity >= DEADLY_CHOP_VELOCITY) entity.health = 0;
+    else entity.damage(VEL_DAMAGE_MODIFIER * y_velocity);
+
+    // Play effects
+    const eyeLocation = entity.eyeLocation;
+    playChopEffects(eyeLocation);
+
+    // Drop head
+    if (entity.isDead()) dropHead(entity, eyeLocation);
+
+    // Delete cooldown
+    await wait(20, 'ticks');
+    cooldowns.delete(entity);
+  }
+}
+
+function dropHead(entity: Entity, loc: Location) {
+  let drop: ItemStack | undefined;
+
+  // Get drop from entity
+  const key = entity.type.toString();
+  if (drops.has(key)) drop = drops.get(key);
+
+  // Get player head drop
+  if (entity instanceof Player) {
+    const player = entity as Player;
+    drop = new ItemStack(Material.PLAYER_HEAD, 1);
+    const meta = drop.itemMeta as SkullMeta;
+    meta.owningPlayer = player;
+    drop.itemMeta = meta;
+  }
+  if (drop) loc.world.dropItem(loc, drop);
+}
+
+async function playChopEffects(location: Location) {
+  new ParticleBuilder(Particle.BLOCK_DUST)
+    .location(location.clone().add(0, 0.5, 0))
+    .data(Material.REDSTONE_BLOCK.createBlockData())
+    .count(20)
+    .spawn();
+}
+
+// Drop blade on piston retract
+registerEvent(BlockPistonRetractEvent, (event) => {
+  const blocks = event.blocks;
+  for (const block of blocks) {
+    const entities = block.location.add(0, 1, 0).getNearbyEntities(1, 1, 1);
+    for (const entity of entities) {
+      if (entity.type === EntityType.ARMOR_STAND) {
+        const armorStand = entity as ArmorStand;
+        dropBlade(armorStand);
+      }
+    }
+  }
+});
+
+// Drop blade on block break
+registerEvent(BlockBreakEvent, (event) => {
+  const entities = event.block.location.add(0, 1, 0).getNearbyEntities(1, 1, 1);
+  for (const entity of entities) {
+    if (entity.type === EntityType.ARMOR_STAND) {
```

Comment:
instanceof ArmorStand may be helpful; you don't need to explicitly cast if you use it.